File: DynamicPrograming/csharp/CountingChange.cs
using System.Collections.Generic;

namespace DynamicProgrammingSolutions;

public static class CountingChange
{
    public static int CountWays(int amount, IReadOnlyList<int> coins)
    {
        var memo = new Dictionary<(int Remaining, int Index), int>();
        return Dfs(amount, 0, coins, memo);
    }

    private static int Dfs(int remaining, int index, IReadOnlyList<int> coins, IDictionary<(int, int), int> memo)
    {
        if (remaining == 0)
        {
            return 1;
        }

        if (index == coins.Count)
        {
            return 0;
        }

        var key = (remaining, index);
        if (memo.TryGetValue(key, out var cached))
        {
            return cached;
        }

        var ways = 0;
        var coin = coins[index];
        for (var qty = 0; qty * coin <= remaining; qty++)
        {
            ways += Dfs(remaining - qty * coin, index + 1, coins, memo);
        }

        memo[key] = ways;
        return ways;
    }
}

File: DynamicPrograming/python/countingChange.py
from functools import lru_cache
from typing import List


def countingChange(amount: int, coins: List[int]) -> int:
    @lru_cache(maxsize=None)
    def dfs(remaining: int, index: int) -> int:
        if remaining == 0:
            return 1
        if index == len(coins):
            return 0

        coin = coins[index]
        ways = 0
        qty = 0
        while qty * coin <= remaining:
            ways += dfs(remaining - qty * coin, index + 1)
            qty += 1
        return ways

    return dfs(amount, 0)
